File: src/spoj/dynamic_programming/M3TILE.c
// http://www.spoj.com/problems/M3TILE/
// Problem: M3TILE - LATGACH3

#include <stdio.h>

#define MAX_LENGTH 31

#define _111 0
#define _011 1
#define _001 2

#define TRUE 1
long MEMOIZED[MAX_LENGTH][3];
int IS_MEMOIZED[MAX_LENGTH][3];

long calculateCombinations(int length, int mask) {
    if(mask == _111 && !length) {
        return 1;
    }
    
    if(length <= 0) {
        return 0;
    }
    
    if(IS_MEMOIZED[length][mask]) {
        return MEMOIZED[length][mask];
    }
    
    switch(mask) {
        case _111:
            MEMOIZED[length][mask] = calculateCombinations(length - 2, _111) + (calculateCombinations(length, _001) << 1);
            IS_MEMOIZED[length][mask] = 1;
            return MEMOIZED[length][mask];
            
        case _011:
            MEMOIZED[length][mask] = calculateCombinations(length - 1, _111) + calculateCombinations(length - 1, _001);
            IS_MEMOIZED[length][mask] = 1;
            return MEMOIZED[length][mask];
            
        case _001:
            MEMOIZED[length][mask] = calculateCombinations(length - 1, _011);
            IS_MEMOIZED[length][mask] = 1;
            return MEMOIZED[length][mask];
    }
}

int main() {
    /*
    printf("%ld\n", calculateCombinations(2, _111));
    printf("%ld\n", calculateCombinations(8, _111));
    printf("%ld\n", calculateCombinations(12, _111));
    printf("%ld\n", calculateCombinations(28, _111));
    printf("%ld\n", calculateCombinations(29, _111));
    printf("%ld\n", calculateCombinations(30, _111));
     */
    
    int length;
    scanf("%d", &length);
    while(length != -1) {
        printf("%ld\n", calculateCombinations(length, _111));
        scanf("%d", &length);
    }
    
    return 0;
}

File: src/spoj/dynamic_programming/M3TILE.java
package spoj.dynamic_programming;

// http://www.spoj.com/problems/M3TILE/
// Problem: M3TILE - LATGACH3

public class M3TILE {

	enum Mask {
		_111, _011, _001
	}

	static long combinations(int length, Mask border) {
		if ((length == 0) && (border == Mask._111)) {
			return 1;
		}

		if (length <= 0) {
			return 0;
		}

		switch (border) {
			case _111 :
				return (2 * combinations(length, Mask._001) /* + combinations(length - 1, Mask._011) */) + combinations(length - 2, Mask._111);

			case _011 :
				return combinations(length - 1, Mask._111) + combinations(length - 1, Mask._001);

			case _001 :
				return combinations(length - 1, Mask._011);

			default :
				throw new RuntimeException();
		}
	}

	public static void main(String[] args) {
		System.out.println(combinations(2, Mask._111));
		System.out.println(combinations(8, Mask._111));
		System.out.println(combinations(12, Mask._111));
		System.out.println();
		System.out.println(combinations(3, Mask._111));
		System.out.println(combinations(5, Mask._111));
		System.out.println(combinations(7, Mask._111));
	}

}
